File: praca-inzynierska-frontend/src/app/app/model/dto/RequiredStaffDto.ts
import { RequiredStaffDayDto } from "./RequiredStaffDayDto";
import { RequiredStaffBackendDto } from "./backend/RequiredStaffBackendDto";

export class RequiredStaffDto {

    constructor(
        public id: number = 0,
        public mondayStaff: RequiredStaffDayDto,
        public tuesdayStaff: RequiredStaffDayDto,
        public wednesdayStaff: RequiredStaffDayDto,
        public thursdayStaff: RequiredStaffDayDto,
        public fridayStaff: RequiredStaffDayDto,
        public saturdayStaff: RequiredStaffDayDto,
        public sundayStaff: RequiredStaffDayDto
    ) {}

    static of(requiredStaff: RequiredStaffBackendDto) {
        return new RequiredStaffDto(
            requiredStaff.id,
            RequiredStaffDayDto.of(requiredStaff.mondayStaff),
            RequiredStaffDayDto.of(requiredStaff.tuesdayStaff),
            RequiredStaffDayDto.of(requiredStaff.wednesdayStaff),
            RequiredStaffDayDto.of(requiredStaff.thursdayStaff),
            RequiredStaffDayDto.of(requiredStaff.fridayStaff),
            RequiredStaffDayDto.of(requiredStaff.saturdayStaff),
            RequiredStaffDayDto.of(requiredStaff.sundayStaff)
        )
    }

    getDayStaff(date: Date): RequiredStaffDayDto {
        switch (date.getDay()) {
            case 1:
                return this.mondayStaff;
            case 2:
                return this.tuesdayStaff;
            case 3:
                return this.wednesdayStaff;
            case 4:
                return this.thursdayStaff;
            case 5:
                return this.fridayStaff;
            case 6:
                return this.saturdayStaff;
            case 0:
                return this.sundayStaff;
            default:
                return undefined;
        }
    }
}

File: praca-inzynierska-backend/application/src/main/kotlin/jwer/pracainzynierskabackend/model/dto/RequiredStaffDto.kt
package jwer.pracainzynierskabackend.model.dto

import jwer.pracainzynierskabackend.utils.TimePointArray
import jwer.pracainzynierskabackend.model.entity.RequiredStaff
import jwer.pracainzynierskabackend.utils.iterator
import java.time.DayOfWeek
import java.time.LocalDate

data class RequiredStaffDto(

        val id: Long,

        val mondayStaff: RequiredStaffDayDto,
        val tuesdayStaff: RequiredStaffDayDto,
        val wednesdayStaff: RequiredStaffDayDto,
        val thursdayStaff: RequiredStaffDayDto,
        val fridayStaff: RequiredStaffDayDto,
        val saturdayStaff: RequiredStaffDayDto,
        val sundayStaff: RequiredStaffDayDto

) {
    constructor(requiredStaff: RequiredStaff)
    : this(
            requiredStaff.id,
            RequiredStaffDayDto(requiredStaff.mondayStaff),
            RequiredStaffDayDto(requiredStaff.tuesdayStaff),
            RequiredStaffDayDto(requiredStaff.wednesdayStaff),
            RequiredStaffDayDto(requiredStaff.thursdayStaff),
            RequiredStaffDayDto(requiredStaff.fridayStaff),
            RequiredStaffDayDto(requiredStaff.saturdayStaff),
            RequiredStaffDayDto(requiredStaff.sundayStaff)
    )

    /**
     * @param start (inclusive)
     * @param finish (inclusive)
     */
    fun toRequiredStaffArray(start: LocalDate, finish: LocalDate): Array<Int> {
        val array = TimePointArray<Int>(start, finish, { it })
        for (date in start..finish) {
            getStaffByDate(date).timePeriods.forEach {
                array[it.timePeriod.atDate(date)] = it.employeeCount
            }
        }
        return array.array
    }

    private fun getStaffByDate(date: LocalDate): RequiredStaffDayDto {
        return when (date.dayOfWeek) {
            DayOfWeek.MONDAY -> mondayStaff
            DayOfWeek.TUESDAY -> tuesdayStaff
            DayOfWeek.WEDNESDAY -> wednesdayStaff
            DayOfWeek.THURSDAY -> thursdayStaff
            DayOfWeek.FRIDAY -> fridayStaff
            DayOfWeek.SATURDAY -> saturdayStaff
            DayOfWeek.SUNDAY -> sundayStaff
        }
    }
}